pyretailscience/cross_shop.py

```diff
@@ -98,35 +112,44 @@ def _calc_cross_shop(
             ValueError: If the groups are not mutually exclusive.
         """
         cols = ColumnHelper()
-        if isinstance(group_1_idx, list):
-            group_1_idx = pd.Series(group_1_idx)
-        if isinstance(group_2_idx, list):
-            group_2_idx = pd.Series(group_2_idx)
-        if group_3_idx is not None and isinstance(group_3_idx, list):
-            group_3_idx = pd.Series(group_3_idx)
-
-        cs_df = df[[cols.customer_id]].copy()
-
-        cs_df["group_1"] = group_1_idx.astype(int)
-        cs_df["group_2"] = group_2_idx.astype(int)
-        group_cols = ["group_1", "group_2"]
 
-        if group_3_idx is not None:
-            cs_df["group_3"] = group_3_idx.astype(int)
-            group_cols += ["group_3"]
+        if isinstance(df, pd.DataFrame):
+            df: ibis.Table = ibis.memtable(df)
+        temp_value_col = "temp_value_col"
+        df = df.mutate(**{temp_value_col: df[value_col]})
 
-        if (cs_df[group_cols].sum(axis=1) > 1).any():
-            raise ValueError("The groups must be mutually exclusive.")
+        group_1 = (df[group_1_col] == group_1_val).cast("int64").name("group_1")
+        group_2 = (df[group_2_col] == group_2_val).cast("int64").name("group_2")
+        group_3 = (df[group_3_col] == group_3_val).cast("int64").name("group_3") if group_3_col else None
 
-        if not any(group_1_idx) or not any(group_2_idx) or (group_3_idx is not None and not any(group_3_idx)):
-            raise ValueError("There must at least one row selected for group_1_idx, group_2_idx, and group_3_idx.")
-
-        cs_df = cs_df.groupby(cols.customer_id)[group_cols].max()
-        cs_df["groups"] = cs_df[group_cols].apply(lambda x: tuple(x), axis=1)
-
-        kpi_df = df.groupby(cols.customer_id)[value_col].agg(agg_func)
-
-        return cs_df.merge(kpi_df, left_index=True, right_index=True)
+        group_cols = ["group_1", "group_2"]
+        select_cols = [df[cols.customer_id], group_1, group_2]
+        if group_3 is not None:
+            group_cols.append("group_3")
+            select_cols.append(group_3)
+
+        cs_df = df.select([*select_cols, df[temp_value_col]]).order_by(cols.customer_id)
+        cs_df = (
+            cs_df.group_by(cols.customer_id)
+            .aggregate(
+                **{col: cs_df[col].max().name(col) for col in group_cols},
+                **{temp_value_col: getattr(cs_df[temp_value_col], agg_func)().name(temp_value_col)},
+            )
+            .order_by(cols.customer_id)
+        )
+        cs_df = cs_df.mutate(
+            groups=ibis.literal("(")
+            + cs_df[group_cols[0]].cast("string")
+            + ", "
+            + cs_df[group_cols[1]].cast("string")
+            + (", " + cs_df[group_cols[2]].cast("string") if group_3 is not None else "")
+            + ibis.literal(")"),
+        ).execute()
+        cs_df["groups"] = cs_df["groups"].apply(ast.literal_eval)
```

Comment:
Can you do the tuple conversion via Pandas please. Also, I think this returns a comma deliminated script of the tuple rather than the tuple?
```suggestion
        cs_df["groups"] = cs_df[group_cols].apply(tuple, axis=1)
        cs_df = cs_df.drop(columns=group_cols)
```